File: 871-keys-and-rooms/keys-and-rooms.c
#include <stdbool.h>
#include <stdlib.h>

void dfs(int** rooms, int* roomsColSize, int roomsSize, int room, bool* visited) {
    if (visited[room]) {
        return;
    }
    
    visited[room] = true;
    
    for (int i = 0; i < roomsColSize[room]; i++) {
        dfs(rooms, roomsColSize, roomsSize, rooms[room][i], visited);
    }
}

bool canVisitAllRooms(int** rooms, int roomsSize, int* roomsColSize) {
    bool* visited = (bool*)calloc(roomsSize, sizeof(bool));
    
    dfs(rooms, roomsColSize, roomsSize, 0, visited);
    
    for (int i = 0; i < roomsSize; i++) {
        if (!visited[i]) {
            free(visited);
            return false;
        }
    }
    
    free(visited);
    return true;
}

File: 871-keys-and-rooms/keys-and-rooms.java
import java.util.*;

class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        Set<Integer> visited = new HashSet<>();
        dfs(rooms, 0, visited);
        return visited.size() == rooms.size();
    }
    
    private void dfs(List<List<Integer>> rooms, int room, Set<Integer> visited) {
        if (visited.contains(room)) {
            return;
        }
        
        visited.add(room);
        
        for (int key : rooms.get(room)) {
            dfs(rooms, key, visited);
        }
    }
}